bump Amlogic to Linux 5.6

chewitt opened the pull request:
This is the main batch of changes I have been hoarding for some time. It's a bit of a "three steps forwards, two steps backwards" update as overall support and packaging for newer devices is nicely improved but there's a large user-experience regression due to the still-ongoing work for hardware decoding and audio in the kernel, ffmpeg and Kodi. There are still some bits missing to make good images, but I need to offload the stable changes and focus efforts onto those three elements.

Notes:

The meson-firmware repo is now obsolete and dropped as vdec firmwares are now upstream and covered with the latest kernel-firmware package bump.

The u-boot patches for VIM3L support are already upstream and can be dropped in the bump to 2020.04. We are now building mainline u-boot for all Amlogic SBC devices. I have reconfigured the FIP sources into a git repo that's better packaged for our use. Switching some of the image bits over to ATF is something to be explored further, but isn't a high priority right now.

The systemd change ensures console access from UART works. I'm not sure when it was dropped or something changed, but without this I have UART output on-screen but no access to the OS to run commands (useful when fiddling with devices off-network).

The `emmctool` helper script is expanded to handle more tasks related to writing LE images to internal eMMC storage. It checks for an explicit list of supported vendors so will not work on random box devices to avoid support issues.

I've added some firmware required for several USB wireless devices (with in-kernel drivers) that I use when testing boxes without Ethernet or working WiFi. Linux 5.6-rc2 brings support for BCM4359 WiFi which is present in many G12B devices.

Linux 5.6-rcX has been quite stable for me as an OS core but (being an rc release) there will be changes to come. I will send updated Linux patches nearer to the 5.6 release date. Hopefully by that point audio support has improved - audio for older GX devices is now partially upstream. I also need to send a pile of minor fixes and device-trees upstream to reduce our patch count.

dhewg commented on projects/Amlogic/bootloader/install:
```diff
@@ -50,7 +114,7 @@ case "${UBOOT_SYSTEM}" in
     cp $FIPDIR/lpddr4_2d.fw $DESTDIR/
     cp $FIPDIR/piei.fw $DESTDIR/
-    cp u-boot.bin $DESTDIR/bl33.bin
+    cp $(get_build_dir $BOOTLOADER)/u-boot.bin $DESTDIR/bl33.bin
```
Looks like you reintroduced that with a rebase, see 7d13537525faf278d0ca2c45e308e3af39b19c38

dhewg commented on projects/Amlogic/bootloader/install:
```diff
@@ -98,13 +162,13 @@ case "${UBOOT_SYSTEM}" in
+    cp $FIPDIR/bl30.bin $DESTDIR/
+    cp $FIPDIR/bl301.bin $DESTDIR/
+    cp $FIPDIR/bl31.img $DESTDIR/
+    cp $(get_build_dir $BOOTLOADER)/u-boot.bin $DESTDIR/bl33.bin
```
same here

dhewg commented on projects/Amlogic/bootloader/install:
```diff
@@ -133,17 +197,17 @@ case "${UBOOT_SYSTEM}" in
+    cp $FIPDIR/bl30.bin $DESTDIR/
+    cp $FIPDIR/bl301.bin $DESTDIR/
+    cp $FIPDIR/bl31.img $DESTDIR/
+    cp $(get_build_dir $BOOTLOADER)/u-boot.bin $DESTDIR/bl33.bin
```
and here

HiassofT commented:
please drop the systemd debug-shell commit. Debug shell / serial console can be enabled with the systemd.debug_shell=ttyXXX kernel command line parameter since commit dd2694fc74ada90ba18037ef6861f872f2b70747 and the rest of the commit just moves enable_service calls around

chewitt commented:
systemd change dropped (moved to boot params) and install script revised.

chewitt commented:
Rebased after #4196 was merged. I also added a busybox change to add the i2cset/get/detect which are useful for confirming i2c things are working on VIM boards (some testing related to fans) and I added more alsa confs. There is alsa a Kodi patch for the Beelink remote shipped with the GT-King box where the airmouse renders the GUI unusable. The patch is a workaround not a fix - I'm still investigating how to handle the device in the long-term so a fix can be upstreamed.

HiassofT commented:
we already ship i2c-tools in the system-tools addon https://github.com/LibreELEC/LibreELEC.tv/blob/master/packages/addons/addon-depends/system-tools-depends/i2c-tools/package.mk - better use that instead of enabling the busybox applet

chewitt commented:
@HiassofT I wasn't aware of it being in system-tools - I've dropped that commit.

dhewg commented:
In case you want i2c-tools in your private builds for testing purposes:
Since #4042 and #4088 it's now possible to include packages of addons in images, without the need to install the addon itself. I to that myself:
```
ADDITIONAL_PACKAGES="diffutils patch screen iw"
```

chewitt commented:
I've added a firmware package for the QCA9377 module on the O2.cz box that I'll be using for some DVB testing/experiments. Kernel patches etc. are updated to 5.6-rc4.

MilhouseVH commented:
Can you add the QCA licence to /licenses, please and I'll merge this.

Minor quibble: although we no longer enforce this requirement, the filenames of the amlogic linux patches should start with `linux` not `amlogic` (ie. start with `$PKG_NAME`).

It's a good idea splitting the patches though - that will pay dividends as commits slowly appear upstream as it's often a huge PITA trying to work out which one commit now needs to be removed from a monolithic patch set.

chewitt commented:
I've moved the qca-firmware repo under the LE github org and added the license text.

CvH commented:
Todo for an follow up PR, Alsa configs need to get combined into 3-4 configs instead of 19 files
`projects/Amlogic/filesystem/usr/share/alsa/cards/`

MilhouseVH commented:
Since this PR (in test build [`#0308`](https://forum.kodi.tv/showthread.php?tid=343069&pid=2930140#pid2930140), we now get the following logged in Kodi:
```
2020-03-29 23:54:02.709 T:1034   ERROR: LoadMappings - ignoring node "iMON HID device" with invalid vendor_product attribute
2020-03-29 23:54:02.709 T:1034   DEBUG: LoadMappings - loaded node "iMON HID device"
```


Full debug log: http://ix.io/2fWg

MilhouseVH commented:
Maybe it's a valid vendor on Amlogic, dunno, but RPi/RPi2/RPi4/Generic now all moan about iMON.

MilhouseVH commented:
I'm guessing this is a typo: `00B2.0003` and the `.` should be a `:`.

HiassofT commented:
My guess would be `2319:00B2` - the change in https://github.com/LibreELEC/LibreELEC.tv/pull/4220/commits/d27bc0748137eecfd026da5d1bd2f8017682362f looks like a hwdb evdev/input match entry (the first 0003 being USB bus followed by vendor+product id and the 0003 at the end being version id)

chewitt commented:
I probably fat-thumbed something, although even if broken it still appears to have the desired effect of stopping the remote from going nuts else I'd have noticed when testing. I'm up to my neck in a work project so might not be able to test a fix tonight or tomorrow.

HiassofT commented:
on closer inspection I can't see how the beelink remote change could have any (positive) effect. Kodi simply ignores vendor:product entries that don't have exactly one colon

see https://github.com/xbmc/xbmc/blob/master/xbmc/peripherals/Peripherals.cpp#L507-L520

my guess is the beelink remote commit can simply be dropped
